### DISCUSSION TITLE
Is it considered bad practive to share a parameter between a dependency and its caller?

### DISCUSSION THREAD

**Discussion by sm-Fifteen** [fastapi/fastapi#8128] [Questions] [answered]
**Description**

One of the features of FastAPI's dependency injection system is grouping parameter sets and the related processing into separate functions and then added to function signatures to create what could very well be a large dependency tree. On occasions, the same parameters could need to be used at multiple levels of the dependency tree, and this creates a problem where the function parameter names *have to match* beween the parent and child.

```py
def square_numbers(numlers: List[int] = Query(...)):
    return [num**2 for num in numlers]

@app.get('/')
def dep_test(numbers: List[int] = Query(...), squared_numbers: List[int] = Depends(square_numbers)):
    return dict(num1=numbers, num2=squared_numbers)
```

The argument `number` is misspelled `numler` in the dependency, which gets added as a separate parameter in the API. While this is a contrived example, FastAPI makes function parameter names unusually important, which could cause problems in cases of parameters that are shared between parent and child like this. 

I'm not too experienced with dependency injection, so maybe I'm just out of the loop regarding the dangers of "intentional collision" like this, but is this generally considered a bad idea or could features be implemented to make this into a feature, like giving dependency parameters an `override`-like flag to make sure the caller shares the same parameter?

**Comment by tiangolo** [comment, marked as answer]
Hmm, that's more or less a "feature".

FastAPI can infer what's the name of the query that you need from the parameter name itself. If you want to use the same parameter in a dependency and the path operation, you can use the same parameter name.

Still, if you want to use a different parameter in your function for some reason but want it to be a different query parameter in your API (or the same you declare somewhere else), you can use `alias` in the `Query()`: https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#alias-parameters

**Reply by sm-Fifteen** [reply]
Marking this as the right answer, as there has been no further objections in 3½ years, and "bad practice" is relative anyway. I feel like there is some issue to raise around this regarding dependencies forcing the use of specific parameters (especially with regards of Query vs Body parameters, where you might want to switch between one or the other), but I don't have a strong enough use case for that at the moment.

**Comment by tiangolo** [comment]
I think the previous comment probably answered your question, so I'll close this issue now. But feel free to add more comments or create new issues.
